fully separable wavelet transform

grlee77 opened the pull request:
This PR implements an alternative multi-level discrete wavelet transform for n-dimensional signals.  The two new functions added are `fswt` and `ifswt`, corresponding to the forward and inverse fully separable wavelet transform.

It differs from `wavedecn` in that it applies all levels of decomposition along a single axis before moving on to the next axis.  This variant has been named the "fully separable wavelet transform", "tensor wavelet transform" or "hyperbolic wavelet transform" and has advantages over the standard isotropic (Mallat) decomposition for anisotropic data.

I wanted to gauge whether there is interest in having this functionality in PyWavelets.  There is a summary of various publications on this technique here:
http://www.laurent-duval.eu/siva-wits-wavelet-anisotropic-non-standard-square-rectangular.html

The images below (with coefficients scaled up for visibility) illustrate the difference (4-level decomposition shown).

![dwt_types](https://cloud.githubusercontent.com/assets/6528957/19195945/93ee9d62-8c81-11e6-971f-fecca8401dc9.png)

TODO:
- [x] add tests
- [x] add example
- [x] return a transform object instead of 2 separate outputs
- [x] extend per-axis wavelet/mode support to fswt, ifswt

grlee77 commented:
If there is interest in this PR, we will also need to decide on the best API for the formatting of the coefficients.  Unlike for the existing `wavedecn`, `waverecn`, it will always be possible to tightly pack the coefficients into a single array.  For now, I return this single n-dimensional array along with a list of slice objects that can be used to retrieve the individual sets of coefficients.

As a concrete example, If a 2D `fswt` has levels = (4, 3), then there will be a total of 4 * 3 subarrays

rgommers commented:
There's a good use case (anisotropic data) and two the papers on this method you linked have hundreds of citations. So common enough to add I think.

rgommers commented:
The list of slices return is slightly ugly imho. Coefficients in a single array is nice, but I wonder if it's possible to make that more self-describing somehow. Otherwise it's overall not really better than the dicts returned from other functions.

rgommers commented:
Related to gh-103, "Indexing the result of multilevel & multidimensional transforms". 

The self-describing bit is something like ``TransformResult`` in gh-93.

grlee77 commented:
There are a few possibilities for API on the coefficients.

**Here is one possibility for a coefficient dictionary:**
``fswt``/``ifswt`` are most closely related in function to ``wavedecn``/``waverecn``.  Let's assume as an example case a full decomposition on an array of dimension 256 x 8.  Here ``wavedecn`` would perform at most 3 levels for a Haar wavelet and each level would correspond to a coefficient dictionary with entries 'ad', 'da', 'dd'.  
However for ``fswt``, the number of levels would be different along each axis: (8, 3).  The transform is first 8 levels of decomposition along axis 0 followed by 3 levels along axis 1 for a total of 8x3=24 sets of coefficients.  It is not possible to cleanly separate these into a list corresponding to "levels" as in ``wavedecn`` anymore.  If we were to use a dictionary, I think the keys would have to be something like a tuple equal in length to the number of axes where (0, 0) would be the approximation coefficients and (1, 3) for example would correspond to one level of decomposition along axis 0 and 3 levels along axis 1.

rgommers commented:
This needs a rebase on master, the ``axis`` fixes to ``dwt``/``idwt`` are already merged.

rgommers commented:
The TODO write tests in the summary can also be checked it looks like.

grlee77 commented:
I have addressed the docstring comments and added the functions to the API documentation.  I still have to revisit the format for returning the coefficients and add an example.

grlee77 commented:
New example of processing an synthetic anisotropic image was added.
Per-axis wavelet/mode support was added.

The only remaining issue is implementing a `TransformResults` object.  I am not sure of the best strategy for that.  For example, if the object stores the wavelets, modes, etc. then you could perform the inverse transform via a simple call.  
```reconstructed = TransformResult.inverse()```
That would be a big departure from how the transforms are currently implemented elsewhere, though.

rgommers commented:
Liking the Mondrian-like image:)

rgommers commented:
>  For example, if the object stores the wavelets, modes, etc. then you could perform the inverse transform via a simple call.

Don't think we considered that before, but it's an interesting idea. The question is then what to do with that object is you perform some operation like thresholding on a ``TransformResult`` instance. That would make the stored info in there incorrect (in that ``inverse()`` won't give the original data) - is that problematic?

grlee77 commented:
> The question is then what to do with that object is you perform some operation like thresholding on a TransformResult instance. That would make the stored info in there incorrect (in that inverse() won't give the original data) - is that problematic?

I don't think that is too problematic.  That is what already happens if you feed thresholded coefficients back into `wavrecn` or any of the other inverse transforms.  The name could be something like `inverse_tranform` rather than `inverse` to help make it clearer that this runs the inverse *transformation* (which should only produce the original data if the coefficients were unmodified).

grlee77 commented:
Hi @rgommers.  I am revisiting this PR after a long absence.

To recap, in your previous review we had decided that some sort of `TransformResult` class should be returned rather than having multiple outputs corresponding to the coefficient array and subband slices.

There is only one new commit here that introduces a `TransformResult` class.  This class also keeps track of which axes, wavelets and edge modes were used so that these no longer need to be passed into the inverse transform.  This makes `ifswt` so that a `TransformResult` is its only argument, meaning that it could just as well be a **method** of the `TransformResult` object rather than a separate function. For consistency with the primarily functional nature of `PyWavelets`, I had left it as a standalone function for now.

rgommers commented on pywt/_multilevel.py:
```diff
@@ -825,3 +828,368 @@ def array_to_coeffs(arr, coeff_slices, output_format='wavedecn'):
+    coeffs : ndarray
+        The coefficient array.
+    coeff_slices : dict
+        Dictionary of slices corresponding to each detail or approximatin
```
typo "approximation"

rgommers commented on pywt/_multilevel.py:
```diff
@@ -825,3 +828,368 @@ def array_to_coeffs(arr, coeff_slices, output_format='wavedecn'):
+        raise ValueError("Axis greater than data dimensions")
+
+
+class TransformResult(object):
```
This name is probably too generic if the class is specific to FSWT. How about something like `FswtResult`?

rgommers commented on pywt/_multilevel.py:
```diff
@@ -825,3 +828,368 @@ def array_to_coeffs(arr, coeff_slices, output_format='wavedecn'):
+
+    Returns
+    -------
+    nD array of reconstructed data.
```
this doesn't render quite right. Maybe this:
```
reconstructed : ndarray
     Array of reconstructed data.
```

rgommers commented:
this needs a rebase now that the `swt2` coefficients PR has been merged

rgommers commented on doc/source/ref/nd-dwt-and-idwt.rst:
```diff
@@ -21,3 +23,15 @@ Multilevel decomposition - ``wavedecn``
+
+Multilevel fully separable reconstruction coeffs - ``TransformResult``
+-----------------------------------------------------------------------
+.. autoclass:: TransformResult
```
this needs renaming too

rgommers replied:
fixed this, updated a test for the `level > dwt_max_level` change, and rebased

rgommers commented on doc/source/conf.py:
```diff
@@ -215,3 +215,7 @@
 exclude_patterns = ['substitutions.rst', ]
+
+# numpydoc_show_class_members = False
+numpydoc_class_members_toctree = False
```
built the docs locally, all looks good. for future reference, do you remember why this was necessary?

grlee77 replied:
I had forgotten, but without it I get dozens of warnings related to the FswtResult class.

See:. https://github.com/numpy/numpydoc/pull/6

grlee77 replied:
I had forgotten, but without it I get dozens of warnings related to the FswtResult class.

See:. https://github.com/numpy/numpydoc/pull/6

rgommers replied:
Ah yes, now I remember. Thanks

grlee77 commented:
Thanks. The Appveyor failure looks like a network connectivity error.

rgommers commented:
The demos weren't running, so I pushed one more fix commit - please check. If CI is green then I think it's good to go.

rgommers commented:
this needs another rebase now

grlee77 commented:
I have done the rebase, but upon further reflection, I think it is possible the names `fswt` and ifswt` could lead to confusion with the stationary wavelet transforms (`swt`, `swt2`, `swtn`, `iswt`, `iswt2`,  `iswtn`).

I am considering renaming these to `fswavedecn`, `fswaverecn` since they are just a fully separable variants of `wavedecn` and `waverecn`. What do you think? It is slightly longer to type, but I think it may be a bit clearer.
The last commit here makes this change. If we merge it, I will need to update #301 and #406 to match before they can be merged.

rgommers commented:
The rename makes sense to me. Okay, let's call this good:) Merging. Thanks @grlee77
